### DISCUSSION TITLE
How can Bonsai support a new third party hardware(camera)?

### DISCUSSION THREAD

**Discussion by jintaozheng88** [bonsai-rx/bonsai#2026] [Q&A]
Hi,
   Have new USB camera which is not on the supporting camera list, understood that a new library has to be developped to support
the camera, how to do this?  We can do this (we have camera SDK and programming skill)...but is there a guide? Or such work has to be done by bonsai software guy?
 
   Thanks!

JT

**Comment by glopesdev** [comment]
@jintaozheng88 the best reference for this would be the [Creating a Package](https://bonsai-rx.org/docs/articles/create-package.html) page in the docs. It needs to be updated in a few places relating to the debugging experience, but the basics should still work. Any feedback on improvements or missing pieces very much welcome!

Below some feedback on the missing bits:
  - **Debugging**: From Bonsai 2.8 extensions are loaded from a different process which unfortunately breaks a little bit the default debugging experience. Fortunately there is an easy workaround currently described in the [README](https://github.com/bonsai-rx/bonsai?tab=readme-ov-file#debugging).
  - **VS Code**: It is now also possible to develop extensions from VS Code, using the [Bonsai.Templates](https://www.nuget.org/packages/Bonsai.Templates) package for `dotnet new`. Recommended pre-requisites are the [C# for Visual Studio Code Extension](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.csharp).
  - **Camera Extension**: Specifically relating to your question about how to integrate a camera, the best approach right now would probably be to look at examples from other camera extensions in Bonsai, for example [Bonsai.Spinnaker.SpinnakerCapture](https://github.com/bonsai-rx/spinnaker/blob/main/Bonsai.Spinnaker/SpinnakerCapture.cs).
